sql/moz-fx-data-shared-prod/udf_js/bootstrap_percentile_ci/udf.sql

```diff
@@ -0,0 +1,337 @@
+/*
+Calculates a confidence interval using an efficient bootstrap sampling technique for a
+given percentile of a histogram.
+
+See:
+- https://mozilla-hub.atlassian.net/browse/DS-2128
+- https://engineering.atspotify.com/2022/03/comparing-quantiles-at-scale-in-online-a-b-testing/
+
+Users must specify an array of percentiles of interest as the first parameter and a
+histogram struct as the second.
+*/
+CREATE OR REPLACE FUNCTION udf_js.bootstrap_percentile_ci(
+  percentiles ARRAY<INT64>,
+  histogram STRUCT<VALUES ARRAY<STRUCT<key FLOAT64, value FLOAT64>>>,
+  metric STRING
+)
+RETURNS ARRAY<
+  STRUCT<
+    metric STRING,
+    statistic STRING,
+    point FLOAT64,
+    lower FLOAT64,
+    upper FLOAT64,
+    parameter STRING
+  >
+> DETERMINISTIC
+LANGUAGE js
+OPTIONS
+  (library = ['gs://moz-fx-data-circleci-tests-bigquery-etl/qbinom.js'])
+AS
+  """
+  function histogramSort(histogram) {
+    histogram.sort(function(a, b) {return parseFloat(a.key) - parseFloat(b.key);});
+  }
+
+  function histogramIsNonZero(histogram) {
+    return histogram.some(bin => parseFloat(bin.value) > 0.0);
+  }
+
+  function validatePercentile(percentile) {
+    if (percentile < 0 || percentile > 100) {
+      throw "percentile must be a value between 0 and 100";
+    }
+  }
+  
+  function getQuantileSamples(percentile, histogram, n_samples) {
+    /*
+    Returns a sorted array of quantile samples from a `histogram` at a target `percentile`.
+    */
+    validatePercentile(percentile);
+
+    /* ensure the histogram is sorted */
+    histogramSort(histogram);
+
+    /* calculate the cumulative number of elements in the histogram */
+    var total = 0;
+    for (var bin of histogram) {
+        total += bin.value;
+        if ( bin.value > 0 ) {
+          /* include this non-empty bin when drawing samples */
+          bin.cumulative_sum = total;
+        } else {
+          /* ignore this empty bin when drawing samples */
+          bin.cumulative_sum = -1;
+        }
+    };
+
+    var samples = [];
+    /* sample the quantile of interest */
+    for ( var i = 0; i < n_samples; i++ ) {
+      /* draw a random variable from the binomial distribution */
+      var x = qbinom(Math.random(), total, parseFloat(percentile) / 100);
```

Comment:
```suggestion
      var x = qbinom(Math.random(), Math.ceil(total), parseFloat(percentile) / 100);
```

`qbinom` doesn't seem to be able to handle floats. When normalizing histograms values usually become floats, so we need to handle this somehow. I'm not sure if simply rounding it to the closest integer is good enough here?

Reply:
This was a great catch. The fix is here: (https://github.com/mozilla/bigquery-etl/pull/3517/commits/e17e136dd53e6ace1a4dff1c1379e8b73ad0b3bc). The quick fix you proposed was most of the solution, but we needed to make some other changes to get the sampling to behave properly. Essentially:
```js
const scale = 1000;
total = Math.ceil(total * scale);
var x = qbinom(Math.random(), total, parseFloat(percentile) / 100) / total;
```
plus a few other changes in how we handle the samples, since they can now take on non-integer values.